### PR TITLE
Correctly set IsLoaded for navigations on derived types

### LINKED ISSUE #19851: DetachedLazyLoadingWarning On Eagerly Fetched Navigation
When doing an `AsNotTracking()` query with an eager fetch, EF *incorrectly* throws a `DetachedLazyLoadingWarning` when accessing the navigation which was eagerly loaded. This appears to happen when the principle entity is in a hierarchy. 

#### Snippet

```
var user = await db.Users
    .AsQueryable()
    .AsNoTracking()
    .Include(x => x.PullRequests)
    .FirstOrDefaultAsync();

var prs = user.PullRequests;    // Throws here
```

*Repro to follow in comment.

### Further technical details
EF Core version: 3.1.1
Database provider: Npgsql & InMemory
Target framework: .NET Core 3.1

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #19851

### CODE REVIEW

File: src/EFCore/Extensions/NavigationExtensions.cs [hunk 1]
```diff
@@ -77,6 +78,33 @@ public static IEntityType GetTargetType([NotNull] this INavigation navigation)
...
+        ///         when a no-tracking query has eagerly loaded this relationship.
+        ///     </para>
+        ///     <para>
+        ///         This type is typically used by database providers (and other extensions). It is generally
+        ///         not used in application code.
+        ///     </para>
+        /// </summary>
+        /// <param name="navigation"> The navigation loaded. </param>
+        /// <param name="entity"> The entity for which the navigation has been loaded. </param>
+        public static void SetIsLoadedWhenNoTracking([NotNull] this INavigation navigation, [NotNull] object entity)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/NavigationExtensions.cs:93]: This should be in .Internal.NavigationExtensions

Comment (ajcvickers) [reply src/EFCore/Extensions/NavigationExtensions.cs:93]: @AndriySvyryd It's used by providers.

Comment (AndriySvyryd) [reply src/EFCore/Extensions/NavigationExtensions.cs:93]: .Infrastructure.NavigationExtensions?

Comment (ajcvickers) [reply src/EFCore/Extensions/NavigationExtensions.cs:93]: Are you worried about it being too visible? I could make it not an extension method.

Comment (AndriySvyryd) [reply src/EFCore/Extensions/NavigationExtensions.cs:93]: Yes, it shouldn't show up in `OnModelCreating`

File: src/EFCore/Extensions/NavigationExtensions.cs [hunk 2]
```diff
@@ -77,6 +78,33 @@ public static IEntityType GetTargetType([NotNull] this INavigation navigation)
...
+        ///         when a no-tracking query has eagerly loaded this relationship.
+        ///     </para>
+        ///     <para>
+        ///         This method is not exposed as a normal extension method since it is typically used by database providers.
+        ///         It is generally not used in application code.
+        ///     </para>
+        /// </summary>
+        /// <param name="navigation"> The navigation loaded. </param>
+        /// <param name="entity"> The entity for which the navigation has been loaded. </param>
+        public static void SetIsLoadedWhenNoTracking([NotNull] INavigation navigation, [NotNull] object entity)
```
Comment (smitpatel) [inline src/EFCore/Extensions/NavigationExtensions.cs:93]: `INavigationBase` :trollface:

**Comment by smitpatel** [comment]
This did not update Cosmos.

**Pull request merged**